### ISSUE TITLE
Build with CGO_ENABLED=0

### ISSUE THREAD

**Issue by silverwind** [go-gitea/gitea#24658]
### Feature Description

We should aim to build with `CGO_ENABLED=0`, which brings benefits:

- The resulting binary is statically linked and can run inside docker `from: scratch` images, which is ideal for security and image size because it only has the binary in the container, nothing else.
- `xgo` can be removed because without CGO, go can cross-compile everything itself.
- When CGO is enabled, go will switch to a suboptimal and toolchain and afaik also produce suboptimal binaries.

What are the remaining dependencies here? Just sqlite, or is there more?

**Comment by wxiaoguang** [comment]
The only blocker is sqlite.

I do not need sqlite, so I can perfectly cross compile Gitea on my macOS and deploy it to my Linux servers.

**Comment by wxiaoguang** [comment]
Are we willing to take the risk of an complex essential core library getting unmaintained?

**Comment by lunny** [comment]
There are two build tags SQLite and pam. Both are optional for build from source. So we are discussing the official release.

Since Gitea with SQLite are welcome by many individuals users. I don't think we should provide the official binaries without SQLite support.

**Comment by lunny** [comment]
> There is https://github.com/glebarez/go-sqlite, not sure how stable it is though, but I guess if it survives a few runs in our CI, it should be good enough for a spin.

It's the same as our PR

**Comment by wxiaoguang** [comment]
The problem is, SQLite is popular doesn't mean its pure-go solution is popular.

Git is much more popular than sqlite, but gogit is far from complete.

**Comment by silverwind** [comment]
Git is also much more complicated than SQLite I assume. Probably by a factor of 10.

**Comment by zeripath** [comment]
I'm just putting a note here to say we already can build with CGO_ENABLED=0 - in fact it's part of our test suite. There are costs though:

* No SQLite
* No PAM

**Comment by wxiaoguang** [comment]
> Git is also much more complicated than SQLite I assume. Probably by a factor of 10.

At least I can see gogit is used by some apps but I doubt whether pure-go-sqlite is battle-tested.

I would still question whether it's worth to use CCGO in Gitea. What if there is an unfixable bug?

* If we use a 3rd go package, even if there is a bug, we can fork and fix.
* But if there is a bug in CCGO related package, how to fix it? I don't have the time to patch the CCGO compiler ....

**Comment by lunny** [comment]
> I'm just putting a note here to say we already can build with CGO_ENABLED=0 - in fact it's part of our test suite. There are costs though:
> 
>     * No SQLite
> 
>     * No PAM

Maybe we can close this issue or we can change the title like `Support sqlite without CGo` ?

**Comment by silverwind** [comment]
What features is PAM used for? At least I don't see it as on option for authentification sources.

**Comment by wxiaoguang** [comment]
* #13257
* #16639 
* #19249 

and more

**Comment by silverwind** [comment]
IDK how PAM integration was done, but it does seem it should be its own authentification source, which could be made unavailable with a specific build tag.

**Comment by lunny** [comment]
In fact, for our current build progress(with xgo), we can build linux official release with the PAM tag which is different from other OS.
But for the future, if we want to remove xgo, we may don't want to do that.

**Comment by p0da** [comment]
Disregarding the PAM issue, which is of limited concern on single tenant systems with only basic auth configured (something that I presume quite a few Gitea instances are) could we look further into this issue? @lunny 

 Not sure how difficult the go-sqlite integration would be but the opt in capacity to run a fully static binary would be very useful, especially in the containerization space.

**Comment by lunny** [comment]
For the default official binary, PAM is not enabled. So the only CGO is from sqlite. Users can build a pure Go binary from source without sqlite support from day 1. But for official binary, we want to support many personal developers/small teams who don't want to install database.

**Comment by p0da** [comment]
@lunny Sorry for the confusion, to clarify I am referring to adding the opt-in capacity to use go-sqlite instead of linked against sqlite through CGO. That is all, this does not have to be shipped in the official binary.

**Comment by lunny** [comment]
To keep the compatible old version, I think making the default official binary without sqlite support will break some Gitea instances.

**Comment by p0da** [comment]
You are likely correct, that is why I previously said it should be an opt-in feature. Advanced users could enabled it at compile time through build tags.

**Comment by lunny** [comment]
Yes. It's ready from day one.  `TAGS="bindata" make build` will generate a Gitea binary without sqlite support. And you can also use `GOOS=linux GOARCH=arm64 TAGS="bindata" make build` for cross-compiling.

Since we cannot remove `sqlite` tag in a short term, maybe we can have a try by zig for cross-compiling with CGO.  https://docs.gitea.com/installation/install-from-source#compile-or-cross-compile-using-linux-with-zig

**Comment by lunny** [comment]
Since #37562 merged, this is possible.

**Comment by silverwind** [comment]
> Since [#37562](https://github.com/go-gitea/gitea/pull/37562) merged, this is possible.

No more blockers like PAM etc?

**Comment by lunny** [comment]
> > Since [#37562](https://github.com/go-gitea/gitea/pull/37562) merged, this is possible.
> 
> No more blockers like PAM etc?

It's not default tags of official releases.

**Comment by bircni** [comment]
This is now resolved by #37562 (Gitea 1.27.0), which makes the pure-Go `modernc` sqlite driver the default. As a result the default build no longer requires CGO, and the Makefile now defaults to `CGO_ENABLED=0` (only enabling CGO for the opt-in `sqlite_mattn` or `pam` tags). SQLite support is retained without CGO, so the goal of building with `CGO_ENABLED=0` is achieved for official builds. Closing as resolved.
